### PR TITLE
Error out during serialization/deserialization in motion node if send/receive functions are not provided

### PR DESCRIPTION
**Pull request by foyzur**
The motion node uses a binary serializer to serialize a type inside SerializeTupleIntoChunks. This binary serializer is supposed to produce at varlena with header to indicate compression info, toasting info etc. The serializer is passed into SerializeTupleIntoChunks as pSerInfo. This pSerInfo originally came during motion layer setup from InitSerTupInfo, which uses CAQL to read the type functions from pg_type.

Unfortunately, if a type doesn't have send/receive, then the InitSerTupInfo forcefully populates out/in functions in send/recv function

If the out function returns only a string, and not a varlena, the headers are not set. Therefore, during byte serialization we will actually obtain a string datum. This datum is then passed to DatumGetByteaP to do the detoasting (if necessary, of course). Unfortunately, pg_detoast_datum cannot find the header info (that was supposed to be in varlena header) as we only have a string datum, and not a varlena. This results in wrong size information and a successive attempt to allocate a very large memory that errors out with memory context error "ERROR:  invalid memory alloc request size". Refer to https://github.com/greenplum-db/gpdb/issues/661.

This PR fixes this by not substituting send/receive with out/in functions. It doesn't, however, error out swiftly (e.g., in parser). This is to ensure highest backward compatibility. In fact, it defers it all the way to actual serialization/deserialization time. It doesn't even error out during InitSerTupInfo() as we may not need a particular column during motion and we may still be successful to execute the query.

### CODE REVIEW

**Comment by foyzur** [comment]
@kuien @HaozhouWang @hardikar @vraghavan78 @d @karthijrk please take a look.

**Comment by danielgustafsson** [comment]
I am a bit worried about throwing the error during execution instead of earlier but I can't propose an approach that doesn't have pitfalls so this is much better than the alternative. The only thing I can think of is throwing a WARNING or NOTICE in CREATE TYPE to highlight this for the user as early as possible.

Either way, we will need to make this requirement very clear in the docs though.

**Comment by hlinnaka** [comment]
We could still use the type input/output functions, if we just convert the cstring to text.

**Comment by hlinnaka** [comment]
FWIW, I don't think it's acceptable to simply disable Motion nodes altogether for types that don't have binary send/receive functions. While you want to have binary send/receive functions for any production-grade datatype, for performance reasons, it's very handy that you can skip them, when developing a new datatype, or just doing quick prototyping.

**Comment by foyzur** [comment]
Hi @hlinnaka I thought about using input/output for just cstring. But, the problem is, how do we detect that the type is just cstring? And users always have option to just mention textsend and textrecv in the type definition for such cases, without having to bother writing a new one. Moreover, even if we perfectly detect that a out/in function is enough, special casing it might be very confusion for the users.

For your other comment about allowing this case during prototyping, my understanding is, the existing code doesn't support such use case. It will expect a varlena and try to extract the header. That's the exact repro of the #661. 

Also, the solution in this PR works with many cases when types have motion but no send/receive functions. This is because of calling SerializeTupleDirect(), instead of SerializeTupleIntoChunks(). In fact, the current solution works with all the ICG tests without breaking any of them. And I had to work extra to come up with a test case that actually goes to SerializeTupleIntoChunks().

**Comment by foyzur** [comment]
Hi @danielgustafsson the idea of an warning sounds really good. I can easily do that as we already had parser level changes and we can just change the ERROR to WARNING. The only concern that I have is, as I mentioned in my previous comment, even without send/receive function, serialization/deserialization works fine with motion if we call SerializeTupleDirect, instead of SerializeTupleIntoChunks. So, in most cases the user may never have to worry. May be doc is the best place to warn the user?

**Comment by hlinnaka** [comment]
On 07/15/2016 08:54 PM, foyzur wrote:
> Also, the solution in this PR works with many cases when types have
> motion but no send/receive functions. This is because of calling
> SerializeTupleDirect(), instead of SerializeTupleIntoChunks(). In
> fact, the current solution works with all the ICG tests without
> breaking any of them. And I had to work extra to come up with a test
> case that actually goes to SerializeTupleIntoChunks().


Oh, wait, there's a whole separate function to handle shorter tuples? 
Why? Why don't we just always do what SerializeTupleDirect does? Why do 
we ever bother calling the send/receive functions?

- Heikki

**Comment by foyzur** [comment]
@hlinnaka my understanding is the network layer (ic_common.c) has a buffer (as expected) that can take it up to a certain size. Without investigating all the details and using my network intuition, it would be like whether you can write your entire tuple using just one packet or you need to chunk it into multiples. Given that network layer is pretty inflexible and not very scalable in terms of buffer size modification, I would assume that's the reason to first try writing directly to the buffer, and if that doesn't succeed, then try chunking (note, chunking would also probably use a ring buffer (I heard they use ring buffer when I asked some network implementation questions to @changleicn) to reassemble those chunks back to a giant tuple). Note, we also use udpifc, which is a modified UDP with flow control, so we probably cannot even depend on TCP stack to do the chunking and flow control for us. Disclaimer: @changleicn or someone else more familiar with udpifc implementation would be better suited to answer this.

**Comment by hlinnaka** [comment]
On 07/15/2016 10:30 PM, foyzur wrote:
> @hlinnaka my understanding is the network layer (ic_common.c) has a
> buffer (as expected) that can take it up to a certain size. Without
> investigating all the details and using my network intuition, it
> would be like whether you can write your entire tuple using just one
> packet or you need to chunk it into multiples. Given that network
> layer is pretty inflexible and not very scalable in terms of buffer
> size modification, I would assume that's the reason to first try
> writing directly to the buffer, and if that doesn't succeed, then try
> chunking (note, chunking would also probably use a ring buffer (I
> heard they use ring buffer when I asked some network implementation
> questions to @changleicn) to reassemble those chunks back to a giant
> tuple).

I get that part, but why do we need the send/receive functions when 
chunking, if we can get away with just writing the raw tuple to the 
wire, when it's small enough? Can't we just write the raw Datums, 
without calling any send/receive/input/output functions?

To put some code where my mouth is, see 
https://github.com/hlinnaka/gpdb/tree/simplify-motion-serialization. 
Basically, I think we should do that.

PS. The regression test needs to be moved to a different file; 
create_type is an upstream test case, so we should keep it unmodified.

- Heikki

**Comment by foyzur** [comment]
Hi @hlinnaka I took a quick look at your changes. It looks sane. I am worried about the disruptive nature of the change though. It's like changing the serialization process at the core. Our strategy at QX team is to always minimize disruptive changes to fix bugs. Features are of course completely different story where we put all our disruptive changes. Do you prefer your solution over this PR one? If so, I can close mine and you can open another PR with your changes. We would also like to push the fix to 4.3 as this is a known bug now. Therefore, please be very careful to test your fix as customer escalations would be pretty bad situation to be at. I would also like the Beijing team @changleicn @yydzero and others, as well as storage team @ashwinstar @asimrp to take a careful look at your PR.

**Comment by hlinnaka** [comment]
Yeah, I definitely think my simpler approach is the way to go. It's a lot simpler, it makes the chunked and non-chunked versions more similar, and it avoids crippling datatypes that have no binary I/O functions.

I don't plan to spend any more time on this myself, so please pick up this from here. I totally agree that it needs more testing. In particular, need to write a test case that tests all the codepaths for pass-by-value and pass-by-ref fixed width datatypes.



On 16 July 2016 03:42:08 EEST, foyzur <notifications@github.com> wrote:
>Hi @hlinnaka I took a quick look at your changes. It looks sane. I am
>worried about the disruptive nature of the change though. It's like
>changing the serialization process at the core. Our strategy at QX team
>is to always minimize disruptive changes to fix bugs. Features are of
>course completely different story where we put all our disruptive
>changes. Do you prefer your solution over this PR one? If so, I can
>close mine and you can open another PR with your changes. We would also
>like to push the fix to 4.3 as this is a known bug now. Therefore,
>please be very careful to test your fix as customer escalations would
>be pretty bad situation to be at. I would also like the Beijing team
>@changleicn @yydzero and others, as well as storage team @ashwinstar
>@asimrp to take a careful look at your PR.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/953#issuecomment-233098417

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by asimrp** [comment]
+1 to Heikki's approach for all the benefits already mentioned in the above comment.  As an aside, I am wondering what is the use of binary input/output functions, if we are not using them for serialization?

Agreed that we should add tests for types having `typbyval = false and typlen > 0`.  Some of these types are already covered by existing tests (e.g. box, circle, money etc).

**Comment by hlinnaka** [comment]
On 17 July 2016 00:16:51 EEST, Asim RP <notifications@github.com> wrote:
>As an aside, I am wondering what is the use of binary
>input/output functions, if we are not using them for serialization?

They're used by binary mode pg_dump, and client connectors that support binary mode. The server itself doesn't need them for anything.

>Agreed that we should add tests for types having `typbyval = false and
>typlen > 0`.  Some of these types are already covered by existing tests
>(e.g. box, circle, money etc).

The codepath in question is only reached when serialaizing a tuple that contains any toasted datums. Otherwise, the fast-path is used. All the existing tests probably take the fast-path. You could use a code coverage tool to check, although it's more straightforward to just add bespoken tests for this.

- Heikki

**Comment by foyzur** [comment]
Closing it for now without merging and will revisit later.

**Comment by hlinnaka** [comment]
Uh, @foyzur, this is a known, live, bug. I hope you intend to revisit this pretty darn soon?
